timesketch/lib/graphs/interface.py

```diff
@@ -0,0 +1,217 @@
+
+from timesketch.lib.datastores.elastic import ElasticsearchDataStore
+
+
+GRAPH_TYPES = {
+    'Graph': nx.Graph,
+    'MultiGraph': nx.MultiGraph,
+    'DiGraph': nx.DiGraph,
+    'MultiDiGraph': nx.MultiDiGraph
+}
+
+
+class Graph(object):
+    """Graph object with helper methods.
+
+    Attributes:
+        nx_instance: Networkx graph object.
+    """
+    def __init__(self, graph_type):
+        """Initialize Graph object.
+
+        Args:
+            graph_type: (str) Name of graph type.
+        """
+        _nx_graph_class = GRAPH_TYPES.get(graph_type)
+        self.nx_instance = _nx_graph_class()
+        self._nodes = {}
+        self._edges = {}
+
+    def add_node(self, label, attributes):
+        """Add node to graph.
+
+        Args:
+            label: (str) Label for the node.
+            attributes: (dict) Attributes to add to node.
+
+        Returns:
+              Instance of Node object.
+        """
+        node = Node(label, attributes)
+        if node.id not in self._nodes:
+            self._nodes[node.id] = node
+        return node
+
+    def add_edge(self, source, target, label, event, attributes=None):
+        """Add edge to graph.
+
+        Args:
+            source: (Node) Node to use as source.
+            target: (Node) Node to use as target.
+            label: (str) Label for the node.
+            event: (dict): Elasticsearch event.
+            attributes: (dict) Attributes to add to node.
+        """
+        edge_id_string = ''.join([source.id, target.id, label]).lower()
+        edge_id = hashlib.md5(edge_id_string.encode('utf-8')).hexdigest()
+
+        edge = Edge(source, target, label, attributes)
+
+        if edge.counter < 500:
+            index = event.get('_index')
+            doc_id = event.get('_id')
+            events = edge.attributes.get('events', {})
+            doc_ids = events.get(index, [])
+            doc_ids.append(doc_id)
+            edge.counter += 1
+            events[index] = doc_ids
+            edge.add_attribute('events', events)
+
+        self._edges[edge_id] = edge
+
+    def commit(self):
+        """Commit all nodes and edges to the networkx graph object."""
+        for node_id, node in self._nodes.items():
+            self.nx_instance.add_node(
+                node_id, label=node.label, **node.attributes)
+
+        for _, edge in self._edges.items():
+            label = edge.label + ' ({0:d})'.format(edge.counter)
+            self.nx_instance.add_edge(
+                edge.source.id, edge.target.id, label=label,
+                **edge.attributes)
+
+    def to_cytoscape(self):
+        """Output graph in Cytoscape JSON format.
+
+        Returns:
+            Graph in Cytoscape JSON format.
+        """
+        cy_json = nx.readwrite.json_graph.cytoscape_data(self.nx_instance)
+        return cy_json.get('elements', [])
+
+
+class BaseGraphElement(object):
+    def __init__(self, label='', attributes=None):
+        self.label = label
```

Comment:
this is a class attribute, please include that in the docstring (same with attributes and id)

Reply:
Fixed